src/client/Client.cc

```diff
@@ -10810,10 +10809,15 @@ void Client::C_Read_Sync_NonBlocking::finish(int r)
     if ((uint64_t)pos >= in->size)
       goto success;
 
+    clnt->put_cap_ref(in, CEPH_CAP_FILE_RD);
     {
       int have_caps2 = 0;
       r = clnt->get_caps(f, CEPH_CAP_FILE_RD, have_caps, &have_caps2, -1);
       if (r < 0) {
+	// Adjust cap_ref - do get_cap_ref again if get_caps fails. Otherwise,
+	// the cap_ref would go negative when C_Read_Finisher::finish_io does
+	// the final put_cap_ref
+        clnt->get_cap_ref(in, CEPH_CAP_FILE_RD);
```

Comment:
Yes you are right. I too thought about it. I think the cap is being given out and requested again because, this involves network call (actual osd call). So to avoid starvation to other requesters, just before network call, it's been dropped and acquired again ?

In objectcacher, the cap is taken only once and they are not dropping and acquring again between multiple reads.

Reply:
> Yes you are right. I too thought about it. I think the cap is being given out and requested again because, this involves network call (actual osd call). So to avoid starvation to other requesters, just before network call, it's been dropped and acquired again ?

Maybe, yes, since put_cap_ref() would flush cap snaps. But I think it suffice here to not put and get again.

Reply:
ok, shall I go ahead and remove put_cap_ref and get_caps here ?

Reply:
yes, please.

Reply:
Done